aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -5,75 +5,247 @@
 
 #include <ATen/core/ivalue.h>
 #include <c10/core/TensorOptions.h>
+#include <ATen/core/Dimname.h>
+
 #include <ATen/core/boxing/KernelFunction.h>
 
-namespace at {
-struct Dimname;
-}
+#include <c10/util/Metaprogramming.h>
 
 namespace c10 {
 namespace impl {
 
+//
+// utils
+//
+
+// is_tensor_ref
+template<class T> struct is_tensor_ref : std::false_type {};
+template <> struct is_tensor_ref<at::Tensor&> : std::true_type {};
```

Comment:
nit: this would be clearer if you say it's a mutable ref. In the C++ standard library, "reference" canonically refers to both const and non-const references, e.g., `static_assert(std::is_reference<const int&>());`

Reply:
definitely